crates/sui-types/src/storage.rs

```diff
@@ -537,6 +562,11 @@ impl InMemoryStore {
         if !self.checkpoints.contains_key(checkpoint.digest()) {
             panic!("store should already contain checkpoint");
         }
+        if let Some(highest_verified_checkpoint) = self.highest_verified_checkpoint {
+            if highest_verified_checkpoint.0 >= checkpoint.sequence_number {
```

Comment:
does this happen in practice? is it preventable, and if so, can we change this to a panic?

Reply:
It does happen in the test I added and can happen in practice too. The scenario is:
```
1. State sync peers announce the lowest checkpoint they have is 10 and highest checkpoint they have is 100
2. Archive has latest data from 0 to 50
3. We start catching up checkpoints from archive from sequence number 0 (and state sync from peers is just waiting) and don't stop until the end (until checkpoint 50)
4. Once the store is caught up to checkpoint 10 (from archive), and is somewhere between checkpoint 10 and 50 - state sync realizes that it can do the catch up from state sync peers at which point there are two parties doing racey read-modify-write on `highest_verified_checkpoint` and `highest_synced_checkpoint` without any synchronization and result is non serial execution. 
```
One obvious way we could prevent this from happening is that catch up from archive never tries to run beyond lowest peer checkpoint (and then it would never overlap with state sync peer based catchup and we would not run into the above race. That means in the above example, catch up from archive stops at checkpoint 10). This is definitely something we could do. The only reason I didn't do it is because download from archive is much faster and if we let catch up from archive (which presumably happens when node is lagging by a lot and can benefit from faster catchups) run all the way till the end, we can have the node catch up quicker.

Reply:
Actually one thing I did not consider before is that how the racing read-modify-write to checkpoint watermark update might interact with checkpoint based tx/fx pruner. It seems like we could get into a state where pruning watermark moved to `X` but state sync from peers ends up inserting tx+fx data for a checkpoint `X-N`. Given this unnecessary complication, I think it might be simplest to have sync-from-archive and sync-from-peers never overlap in the checkpoint range by making sync-from-archive stop at lowest peer checkpoint. So, in the above scenario now:
```
1. State sync peers announce the lowest checkpoint they have is 10 and highest checkpoint they have is 100
2. Archive has checkpoint data from 0 to 50
3. We start catching up checkpoints from archive from sequence number 0 (and state sync from peers is just waiting) and stop catching up once we reach checkpoint 9
4. State sync from peers realizes that it can now do the catch up from peers and starts catching up from checkpoint 10

Reply:
Another scenario we could still have both state-sync-from-archive and state-sync-from-peers still overlap is something like this:
```
1.  State sync peers announce the lowest checkpoint they have is 10 and highest checkpoint they have is 100. The lowest_peer_checkpoint_watermark is 10
2. Archive has checkpoint data from 0 to 50
3. We start catching up checkpoints from archive from sequence number 0 (and state sync from peers is just waiting) until checkpoint 10 (checkpoint 9)
4. A new state sync peer joins or an existing peer backfills its own state such that lowest_peer_checkpoint_watermark moves back to 5
5. In such a scenario, state sync from peers will also start trying to download checkpoints in parallel and both tasks could race trying to perform read-modify-write on the checkpoint watermarks

Reply:
could we just have a per-checkpoint mutex?

Reply:
Just want to make sure I understand the suggestion - instead of having one mutex for the watermark, we could instead have a mutex per checkpoint. So, before inserting checkpoint data in the db we will first read from db under a mutex whether the checkpoint already exists? It seems more expensive overall to have a mutex per checkpoint in memory and doing an extra db lookup operation before insert

Reply:
i think a mutex for the watermark would be fine too